Sources/SwiftSyntax/SyntaxNodes.swift.gyb.template

```diff
@@ -155,6 +155,33 @@ public struct ${node.name}: ${base_type}Protocol, SyntaxHashable {
     return ${node.name}(newData)
   }
 %     end
+
+%     # ===================
+%     # Validate layout
+%     # ===================
+
+  public func _validateLayout() {
+%     if node.is_unknown():
+    // We are verifying an unknown node. Since we don’t know anything about it
+    // we need to assume it’s valid.
+%     else:
+    let rawChildren = Array(RawSyntaxChildren(Syntax(self)))
+    assert(rawChildren.count == ${len(node.children)})
+%       for i, child in enumerate(node.children):
+    // Check ${i}-th child is ${child.type_name} ${"or missing" if child.is_optional else ""}
+%         if not child.is_optional:
+    assert(rawChildren[${i}].raw != nil)
+%         end
+    if let raw = rawChildren[${i}].raw {
+      let info = rawChildren[${i}].syntaxInfo
+      let absoluteRaw = AbsoluteRawSyntax(raw: raw, info: info)
+      let syntaxData = SyntaxData(absoluteRaw, parent: Syntax(self))
+      let syntaxChild = Syntax(syntaxData)
+      assert(syntaxChild.is(${child.type_name}.self))
```

Comment:
Could you speed this up by just checking the kind instead of constructing a SyntaxData and going through the runtime typecheck?

Reply:
We can’t. If the layout requires a `Expr` node we need to make sure that we accept `IdentifierExpr`. And kind doesn’t capture that inheritance relation.

Reply:
Right, good point. Okay, sounds good then.